bgpd/bgp_bmp.c

```diff
@@ -1213,18 +1217,21 @@ static bool bmp_wrqueue_locrib(struct bmp* bmp, struct pullwr* pullwr) {
 		zlog_info("bmp: loc rib monitoring on");
 		struct bgp_path_info *bpi;
 
-		// TODO BMP_MON_LOC_RIB understand this part more, why iterate through the table ?
 		for (bpi = bn ? bgp_dest_get_bgp_path_info(bn) : NULL; bpi;
 		     bpi = bpi->next) {
-//			if (!CHECK_FLAG(bpi->flags, BGP_PATH_VALID))
-//				continue;
+			if (!CHECK_FLAG(bpi->flags, BGP_PATH_SELECTED))
+				continue;
 			if (bpi->peer == peer)
 				break;
 		}
 
-		bmp_monitor(bmp, peer, 0, BMP_PEER_TYPE_LOC_RIB_INSTANCE, &bqe->p, prd,
-			    bpi ? bpi->attr : NULL, afi, safi,
-			    bpi ? bpi->uptime : monotime(NULL));
+		uint8_t peer_distinguisher[8] = {0};
+		if (bmp->targets->bgp->inst_type != VRF_DEFAULT) {
+			memcpy(peer_distinguisher, &bmp->targets->bgp->vrf_id, sizeof(vrf_id_t));
```

Comment:
warning: vrf may not be initialized, and you may have vrf_id  = RF_UNKNOWN.
The below config has a network selected in loc-rib, despite no vrf has been created.
I think BMP should not provision it

```
router bgp 44 vrf 33
 no bgp network import-check
 !
 address-family ipv4 unicast
  network 10.100.0.0/24
 exit-address-family
 !
 bmp targets AAA
 exit
!
line vty
!
end
ubuntu2004(config-router-af)# do show bgp vrf 33 ipv4
BGP table version is 1, local router ID is 0.0.0.0, vrf id -
Default local pref 100, local AS 44
Status codes:  s suppressed, d damped, h history, * valid, > best, = multipath,
               i internal, r RIB-failure, S Stale, R Removed
Nexthop codes: @NNN nexthop's vrf id, < announce-nh-self
Origin codes:  i - IGP, e - EGP, ? - incomplete
RPKI validation codes: V valid, I invalid, N Not found

   Network          Next Hop            Metric LocPrf Weight Path
*> 10.100.0.0/24    0.0.0.0                  0         32768 i

Displayed  1 routes and 1 total paths
```

Reply:
added the fix in [dd6bf8a](https://github.com/FRRouting/frr/commit/dd6bf8ade441007923e83f078cbdedbad5a273ff) too, we'll keep it if u agree

Reply:
having vrf_id = VRF_UNKNOWN and no RD set, we could not completely identify the route_distinguisher. With what is proposed, 0 is returned, whereas I think we should cancel the processing.

a RD may be defined for a given BGP instance, even for the default one.
So I would propose the following:

```
bmp_get_peer_distinguisher()
	if BGP_VPN_POLICY_TOVPN_RD_SET
		use RD
	else if vrf_id != VRF_UNKNOWN
		use vrf_id
	else
		should not be sent
```

Reply:
sorry, just saw this comment.
you're right, with what i proposed we can't differentiate between the VRF_DEFAULT case and the vrf_id == VRF_UNKNOWN + no RD case
i will make a fix in the following weeks for this concern

Reply:
fixed in https://github.com/FRRouting/frr/commit/1b6b89fcc5c0a9222a38b22c60f8646a2b1f7016